### ISSUE TITLE
docs404.py failing

### ISSUE THREAD

**Issue by nmelehan** [linode/docs#1915]
These issues are fixed now, but we have a summary to preserve what happened:
---

The docs404.sh script saw a number of recent interrelated issues that took a few pull requests to resolve. Here's a timeline:

### Prologue
On June 15, we observed that the Thank You page was failing the docs404 test. This was addressed in these PRs:

https://github.com/linode/docs/pull/1849
https://github.com/linode/docs/pull/1852
https://github.com/linode/docs/pull/1855
https://github.com/linode/docs/pull/1858

### New issues on July 12

Starting July 12, @Guaris observed CI tests started to result in false-positives from Travis; in other words, links that were 404s were not reporting as such. Here is a PR from that date whose docs404 test passed, but which reported the existence of 404s in the test logs:

https://github.com/linode/docs/pull/1899
https://travis-ci.org/linode/docs/jobs/403099290
```
404 response in HTML - see logs
No 404s were found from manual test
The command "./ci/scripts/$TEST_SUITE.sh" exited with 0.
Done. Your build exited with 0.
```

Later on July 12, @leslitagordita pushed a PR that fixed a broken image link in the Rocketchat guide: https://github.com/linode/docs/pull/1900

This PR was never merged, but the Travis logs show that no 404 links were found from Scrapy:
```
Scraper did not find any 404 links
No 404s were found from manual test
```

### Continuing issues on July 13

@Jskobos discovered the reason for the false-positives: in [PR 1855](https://github.com/linode/docs/pull/1855/files), we added another test and a second line to docs404.sh:

```
python ci/docs404.py
+python ci/test_urls.py
```

The bash script was judging its success by the result of the last line, which was `python ci/test_urls.py`. The fix was to combine the two lines into one: 

```
python ci/docs404.py && python ci/test_urls.py
```

This was packaged into [PR 1902](https://github.com/linode/docs/pull/1902), along with the update to the Rocketchat guide/image link that @leslitagordita also made in [PR 1900](https://github.com/linode/docs/pull/1900).

### Continuing issues on July 18

[PR 1902](https://github.com/linode/docs/pull/1902) was not merged until July 17. On July 18, @Guaris noticed that PRs were failing the docs404 test again. We found that this was because a dependency of Scrapy called Twisted was failing. The following errors were appearing in the docs404 logs in all PRs from July 13 onward:

```
Traceback (most recent call last):
  File "ci/docs404.py", line 4, in <module>
    import scrapy
  File "/home/travis/virtualenv/python3.6.3/lib/python3.6/site-packages/scrapy/__init__.py", line 34, in <module>
    from scrapy.spiders import Spider

# [...]

    from twisted.web import http
  File "/home/travis/virtualenv/python3.6.3/lib/python3.6/site-packages/twisted/web/http.py", line 100, in <module>
    from twisted.internet import interfaces, protocol, address
  File "/home/travis/virtualenv/python3.6.3/lib/python3.6/site-packages/twisted/internet/address.py", line 103, in <module>
    class UNIXAddress(object):
  File "/home/travis/virtualenv/python3.6.3/lib/python3.6/site-packages/twisted/internet/address.py", line 111, in UNIXAddress
    name = attr.ib(converter=attr.converters.optional(_asFilesystemBytes))
TypeError: attrib() got an unexpected keyword argument 'converter'
```

We found that Twisted had released a new version [on July 13](https://pypi.org/project/Twisted/#history): 18.7.0. In other words, this issue came up simultaneously and coincidentally with our other issues, but was masked by the fact that we had not merged [PR 1902](https://github.com/linode/docs/pull/1902) for a few days after it was created.

The solution was to pin Twisted to version 18.4.0, which was enough to satisfy the Scrapy requirements (which needs a version `>=13.1.0`): https://github.com/linode/docs/pull/1913

We also pinned Scrapy to `1.5.1` in case later versions need a newer version of Twisted:
https://github.com/linode/docs/pull/1914

---

### Conclusion

As of [PR 1914](https://github.com/linode/docs/pull/1914), docs404.sh is no longer generating false-positives nor always failing.

### Other takeaways

At different points in this troubleshooting, we found that the docs404 test was succeeding normally on our MacBooks. I believe this is because when running the tests locally, the local environment already satisfied the pip requirements (from previous invocations). 

In other words, the good and sufficient version of Twisted (18.4.0) was already installed and not being updated to 18.7.0. **We may want to consider force-updating the required packages when testing locally to identify these issues, as that more closely resembles the Travis environment.**

**Comment by GuessWhoSamFoo** [comment]
@nmelehan  The issue is improper handling of exit codes from Python exceptions which remains unresolved

**Comment by nmelehan** [comment]
@GuessWhoSamFoo Hi sam, can i ask you to elaborate a bit on that? The exception [here](https://travis-ci.org/linode/docs/jobs/405362771), for example, resulted in an exit code of 1 for the script, causing the Travis test to fail
